### PR TITLE
Active Learning

### PR DESCRIPTION
**Pull request by y0ast**
A UB code style version of the AL notebook. There are several small open todos in the code, but for the majority we (Andreas+me) thought it's time to share and obtain some feedback on the approach.

This script obtains ~80-90% accuracy with ~150 CIFAR-10 samples.

Open todo:
- ~~A test following `deterministic_test.py`.~~
- ~~Follow `pylint` style from the repo~~
- ~~Add docstrings~~

Left for follow up PR: Proper LR schedule and early stopping - this will improve performance significantly.

Joint work with @BlackHC (Andreas).

### CODE REVIEW

**Comment by BlackHC** [comment]
@googlebot I consent.

🎉

File: baselines/jft/active_learning.py [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+    num_epochs=1,  # Don't repeat
+  )
+
+  test_accuracies = []
+  training_size = []
+
+  rng, rng_loop = jax.random.split(rng, 2)
+  rngs_loop = flax_utils.replicate(rng_loop)
+
+  while True:
```
Comment (dusenberrymw) [inline baselines/jft/active_learning.py:375]: Updating to `while current_train_ds_length < 150` would be easier to quickly understand.

File: baselines/jft/al_utils.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+# on the tf.data.Dataset would work too.
+
+from typing import Dict, List, Optional, Union
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from tensorflow_datasets.image_classification import cifar
+
+
+def subset_generator(dataset, subset_ids):
+    def inner():
```
Comment (dusenberrymw) [inline baselines/jft/al_utils.py:17]: style nit: 2-space indentation.

File: baselines/jft/active_learning.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+    map(lr_fn, range(total_steps)), prefetch_to_device
+  )
+
+  for step, train_batch, lr_repl in zip(tqdm.trange(1, total_steps + 1), iter_ds, lr_iter):
+    opt_repl, l, rngs_loop, measurements = update_fn(opt_repl, lr_repl, train_batch["image"], train_batch["labels"], rngs_loop)
+
+  return opt_repl, rngs_loop
+
+
+def make_init(model, test_ds, local_batch_size, config):
```
Comment (dusenberrymw) [inline baselines/jft/active_learning.py:152]: nit: for consistency with the other `make_*` functions, it might be nice to name this `make_init_fn`.

Comment (dusenberrymw) [inline baselines/jft/active_learning.py:152]: I would recommend replacing `test_ds` with `image_shape` to help decouple this abstraction from the dataset replication details. Then `dummy_input = jnp.zeros((local_batch_size,) + image_shape, jnp.float32)`

File: baselines/jft/al_utils.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+# Custom CIFAR-10 dataset that returns a subset for training
```
Comment (dusenberrymw) [inline baselines/jft/al_utils.py:1]: style nit: place with a triple-quoted section.

File: baselines/jft/active_learning.py [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+    # init head kernel to all zeros for fine-tuning
+    if config.get('model_init'):
+      params['head']['kernel'] = jnp.full_like(params['head']['kernel'], 0)
+
+    return params
+
+  return init
+
+
+def make_update_fn(model, config):
```
Comment (dusenberrymw) [inline baselines/jft/active_learning.py:173]: Thanks for adding these abstractions.

**Review by dusenberrymw — CHANGES_REQUESTED** [review]
Thanks for creating this! Left some comments. I would also strongly recommend adding script-level testing like you mentioned in the open TODO list so that we can iterate while keeping the script deterministically unchanged (when we want it to).

File: baselines/jft/active_learning.py (hunk 1 above)
Comment (BlackHC) [reply baselines/jft/active_learning.py:375]: What's our target Python version at Google? Can we use the `:=` syntax yet?

IIRC the reason for this was just so that `current_train_ds_length` is only assigned once

Comment (dusenberrymw) [reply baselines/jft/active_learning.py:375]: We should target 3.7 compatibility, so can't use `:=` yet.

**Comment by y0ast** [comment]
@dusenberrymw This should now have addressed your comments, contain docstrings and tests.

**Review by dusenberrymw — APPROVED** [review]
Thanks, Joost & Andreas! This LGTM. I'll pull it in and merge.

Three comments/questions:
- Have you both singed the CLA?
- Can you squash the commits to a single commit?
- Also, I believe the commit message for that single commit can be edited such that GitHub will recognize two co-authors if you're interested in having both of you attributed to the commit. I've never done it, but here's [the guide](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors#creating-co-authored-commits-on-the-command-line).

**Comment by y0ast** [comment]
1. yes, CLA is signed by Andreas/Me (see also: https://github.com/google/uncertainty-baselines/pull/610/checks?check_run_id=4370587110)
2. Merged in latest main + squashed
3. Added Andreas as `Co-authored-by`

**Review by dusenberrymw — APPROVED** [review]
